symtabAPI/src/Object-elf.C

```diff
@@ -3646,9 +3646,15 @@ LineInformation* Object::parseLineInfoForObject(StringTablePtr strings)
 
     int status;
 
-    while ((status = dwarf_next_lines(dbg, off = next_off, &next_off, &cu,
-			     &files, &fileCount, &lineBuffer, &lineCount)) == 0)
-    {
+    while (1) {
```

Comment:
Explicit infinite loops like this are technically UB in C++. Most compilers just "do the right thing". I'll make a note to revisit.

Reply:
If you look at the code, all I wanted to do was to enclose the call to dwarf_next_lines in the critical section. Rather than having the test in the loop header, I had the test to break from the loop a few lines below the loop header.

The loop could be for(;;) rather than while (1) if you prefer.

Reply:
>If you look at the code, all I wanted to do was to enclose the call to dwarf_next_lines in the critical section. Rather than having the test in the loop header, I had the test to break from the loop a few lines below the loop header.

Yup. I was originally going to recommend moving the synchronized call to `dwarf_next_line` into a lambda and then put the test back into the `while`. Something like

    auto next_line = [&](){
      return dwarf_next_lines(dbg, off = next_off, &next_off, &cu,
			       &files, &fileCount, &lineBuffer, &lineCount) == 0;
    };

    while(next_line()) {
      ...

but I couldn't prove to myself that was thread-safe wrt the variables used in the dwarf call without looking at the context of the dwarf call in greater detail. I didn't want to hold up this change because you need it soon.

>The loop could be for(;;) rather than while (1) if you prefer.

Those fall under the same rule (intro.progress).

I double-checked the standard and the current form is actually thread-safe so long as OpenMP is enabled because of

```perform a synchronization operation or an atomic operation```

I agree that we should still move to reader/writer semantics so we don't get bad reads or ABBA issues. I'll make a Github issue for us to revisit this.

Reply:
The loop has side effects and an exit condition. I don't think it meets the definition of an infinite loop. I thought the way I rewrote

while (status = dwarf_next_lines(...)) {
...
}

was much better than the following:

as
#pragma omp critical (next_lines)
status = dwarf_next_lines(...)
while (status) {
...
#pragma omp critical (next_lines)
status = dwarf_next_lines(...)
}